product_docs/docs/pgd/5/admin-tpa/upgrading_major_rolling.mdx

```diff
@@ -0,0 +1,537 @@
+  - Copy the current version's configuration files to the new data directory: `postgresql.conf`, `postgresql.auto.conf`, `pg_hba.conf`, and the whole `conf.d` directory.
+1. Rename the PGDATA directories to switch the old versions out and the new versions in.
+  - Move `/opt/postgres/data` to `/opt/postgres/dataold` and move `/opt/postgres/datanew` to `/opt/postgres/data`.
+1. Double-check that both Postgres instances aren't running.
+  - Use `pg_ctl` to stop both `/opt/postgres/dataold` and `/opt/postgres/data`.
+1. Check that you can upgrade safely.
+  - Use the `--check` option when running `bdr_pg_upgrade`.
+1. Run the Postgres major version upgrade.
+  - Run `bdr_pg_upgrade` again without the `--check` option.
+1. Update the Postgres service configuration to the new version and refresh the system's service manager to apply the changes.
+  - Replace the old version number with the new version number in the `postgres.service` file.
+    - `sudo sed -i -e 's/<old_version_number>/<new_version_number>/g' /etc/systemd/system/postgres.service`
+1. Start Postgres.
+  - `systemctl start postgres`
+1. Validate the new Postgres version. 
+  - Connect to the `bdrdb` database and run `SELECT * FROM bdr.group_versions_details;`.
+1. Test the upgrade directly against the newly upgraded node.
+  - Run your test suite again while still connected by way of SSH into the newly upgraded node.
+1. Clean up post-upgrade. 
+  - Run the new PG version's `vacuumdb`.
+  - Remove the old version's data directory, `/opt/postgres/dataold`.
+
+### Reconcile the upgrade with TPA
+
+TPA needs to continue to manage the deployment effectively after the upgrade. Therefore, it's necessary to reconcile the upgraded nodes with TPA. Follow these steps to update the configuration and redeploy the PGD cluster through TPA.
+
+- Change the `config.yml` of the TPA-managed cluster to the new version.
+  - `cluster_vars: postgres_version: '<new_version_number>'`
+- Use `tpaexec` to redeploy the PGD cluster with the updated `config.yml`.
+  - `tpaexec deploy <cluster_name>`
+
+Details of these steps are shown in the worked example, specifically for upgrading a [shadow node](https://www.enterprisedb.com/docs/pgd/latest/terminology/#write-leader) from Postgres 15 to Postgres 16. 
+
+## Worked example
+
+This worked example starts with a [TPA-managed PGD cluster deployed using the [AWS Quickstart](https://www.enterprisedb.com/docs/pgd/latest/quickstart/quick_start_aws/). The cluster has three nodes: kaboom, kaolin, and kaftan, all running Postges 15.
+
+In this scenario, the current write leader is kaftan. This example upgrades the first of two shadow nodes, kaboom.
+
+!!! Note
+Some steps of this process involve running commands as the Postgres owner. We refer to this user as postgres throughout, when appropriate. If you're running EDB Postgres Advanced Server, substitute the postgres user with enterprisedb in all relevant commands.
+!!!
+
+### Confirm the current Postgres version
```

Comment:
Should this not be in the plan your upgrade?

Reply:
I see how it makes sense to put it in the plan. I had thought it would be the first step of the node-by-node steps, such that we would check the versions each time we were on a node to ensure things were being updated as we expected. Let me see if I can balance both these goals.